### DISCUSSION TITLE
Landing page vs. other docs

### DISCUSSION THREAD

**Discussion by lprimak** [docToolchain/docToolchain#1099] [Q&A]
II am still trying to understand the basics... do bear with me :)
Looks like there are two trees under `src/`: `site` and `docs`
Looks like `docs` has the arc42 template a and `site` has the landing page and groovy templates.

I can't find where `arc42` in the right hand corner of the landing page comes from.
And generally, why is the landing page written in html and the other docs in asciidoc?
I would think the landing page would be written in asciidoc? and be in the same source tree?
Is there something fundamental I don't understand?

**Comment by lprimak** [comment]
Looks like there are two disconnected templates, groovy (with html landing page) and arc42.
Why are they so disconnected? Where are the pieces that glue them together?

**Comment by rdmueller** [comment]
the `site` folder is only part of the code, not of your repository where you use docToolchain. It contains the theme with CSS, Javascript and templates for rendering HTML. You only need this if you want to modify the theme/style.

It also includes a default landing page. This is just so that there is a landing page when you start with your project. It will be overwritten in your own repository.

The landingpage is designed as `.gsp` which - in this case - is basically HTML. This is because the landing page should not contain too much text but should be styled in a way that it looks good and appealing. It should describe your product.

The arc42 template is hosted in another repository. You can fetch it through `./dtcw downloadTemplate`. It will ask you for the language and style (with help or without).

Try to follow the first tutorials here: https://doctoolchain.org/docToolchain/v2.0.x/020_tutorial/010_Install.html
It would be great to get feedback on what is missing or hard to understand from these tutorials, so that we can make them better :-)

**Reply by lprimak** [reply]
> It also includes a default landing page. This is just so that there is a landing page when you start with your project. It will be overwritten in your own repository.

I am not sure what you mean by that. Is there documentation how to "overwrite in my own repository?"

**Reply by lprimak** [reply]
I think I understand that one now as well. You can just have one file (i.e. landingpage.jsp) and it will override the default one.
I used `copyTheme` which copied many many files and thought that it's 'all-or-none' type deal.

This is definitely missing from the tutorial and very much non-obvious from the docs themselves

**Reply by rdmueller** [reply]
> I used copyTheme which copied many many files and thought that it's 'all-or-none' type deal.

it first copies everything to your project but you can delete everything which you don't modify.

The mechanism behind this is the following:

- docToolchain creates a rowking directory `build/microsite/tmp` where it first copies the internal theme (the default)
- it then check if a DTC_SITETHEME is defined. This can be configured in `dtcw` and should point to a zipped `site`-folder. If it is defined, it fetches it and copies it over the `build/microsite/tmp/src/site` folder 
- it then copies the site folder from the repository over the `build/microsite/tmp/src/site` folder.

yes, we have to update the docs and add a tutorial for this! Maybe @Frank-dev20 can help here?

**Comment by mh182** [comment]
> II am still trying to understand the basics... do bear with me :) Looks like there are two trees under `src/`: `site` and `docs` Looks like `docs` has the arc42 template a and `site` has the landing page and groovy templates.
> 

We really should think about splitting the current docToolchain repository into two separate repositories: a repository with the code and one with the documtation.

The advantage would be that we would have a template for people to see how they could structure their documentation without exposing them to the docToolchain code (which most people would not be interested in).

On the other side: the documentation is part of the project and it would be painful for docToolchain development.

**Reply by rdmueller** [reply]
I would like to extract the default theme into another repository as example how you can create your own themes.
Regarding a template project - there are some tempalte/demo projects within the docToolchain organisation. I guess we should reference them somewhere. For instance, there is a CI/CD project which shows how to use github actions or gitlab pipelines to render the result.

**Comment by lprimak** [comment]
I think I am starting to better understand what's going on here.
It looks like the reason why the landing page being different from the rest of the docs (in a different tree, written in gsp/html and not asciidoc) is because of JBake legacy.

JBake was originally envisioned for blog sites, which typically use auto-generated index pages, which is perfect for a template.
However, later JBake evolved as a generic SSG platform, but the legacy remained.
In order to turn off this 'legacy', `jbake.properties` file needs an `render.index=false` entry, which permits using 'regular' asciidoc files as index page.

Even after going through the tutorial 3 times, I feel there is lots of extra, unnecessary cognitive load spent because of the following:
- Figuring out the components (docs/ site/ Config.groovy), JBake and how they all fit together
- Figuring out how gradle, dtcw.sh, and doctoolchain.sh fit together
- How the JBake magic works (what's done outside JBake)
- How the components are structures (gradle scripts, shell scripts, menu.groovy etc)

The common theme is that the bits of code are organized outside of their intuitive places where I would expect to find them.

Here are my suggestions:
- Consolidate all Groovy code into a single project into a well-organized submodules, which are named for the exact functionality they provide.
- Move landing page into asciidoc in the `docs/` directory, and _only_ have one place where there is content, i.e. completely separate code from template, and template from the code.
- Remove all code and magic from gradle and into the module above.
- Have *one* configuration file. Currently there is `Config.groovy`, `config.adoc` `configFragment.groovy` and `.asciidocconfig.adoc`, or have clear documentation what each of them do and how they interrelate.

I hope my understanding of this is correct. What do you think?

**Reply by rdmueller** [reply]
This is great feedback for the documentation!

regarding your thoughts about the landingpage: it is intended that the landing page is not just asciidoc but html. The goal is that a user will first see the landing page which is the store front of the project. By using plain html, teams should get the freedom to style it in an appealing way. However, I am also aware that for most dev teams, this is not an easy task.

**Reply by lprimak** [reply]
I would think most projects would already have a landing page and wouldn't need another one. Those existing landing pages would already have an architecture that they are generated in.

In my case, there is already a jbake-based process in place for the landing page and it needs to integrate with what docToolchain does, so the whole website generation is a cohesive, seamless process.

This is why docToolchain itself needs to have an asciidoc landing page that's consistent with everything else docToolchain does, and not be an outlier.

Basically, now I have to migrate the landing page functionality and rework it to my existing processes.
It would have been much, much easier if the landing page was in asciidoc.

**Comment by lprimak** [comment]
And maybe another one, not make users learn gradle :)

**Reply by rdmueller** [reply]
yes. the wrapper already removed gradle from within the users project repository and there will be further changes to hide gradle even more.

**Comment by lprimak** [comment]
Another point of complexity is understanding what the components are and how the relate.
My suggestion would be to clearly separate value-adds from where the integration points are.

In other words, there are many things that docToolchain provides:
- Integrations with other tools (gradle, jbake, template, asciidoc etc)
- Value add (chapters, search, etc)

There should be clear separation of those so value-adds can be used outside (or with different) integrations.

Example: Integrate with maven-based build system using jbake-maven-plugin for other pages already.
